### DISCUSSION TITLE
where to add oauth scopes

### DISCUSSION THREAD

**Discussion by HosMercury** [maxcountryman/axum-login#165] [Q&A] [answered]
I was using oauth2 crate and i was able to add scopes // google oauth 

here trying 
but get client mismatch type
```
        let client_id = env::var("CLIENT_ID")
            .map(ClientId::new)
            .expect("CLIENT_ID should be provided.");
        let client_secret = env::var("CLIENT_SECRET")
            .map(ClientSecret::new)
            .expect("CLIENT_SECRET should be provided");

        let scope_profile: &str = "https://www.googleapis.com/auth/userinfo.profile";
        let scope_email: &str = "https://www.googleapis.com/auth/userinfo.email";

        let auth_url = AuthUrl::new("https://accounts.google.com/o/oauth2/v2/auth".to_string())?;
        let token_url = TokenUrl::new("https://www.googleapis.com/oauth2/v3/token".to_string())?;
        let client = BasicClient::new(client_id, Some(client_secret), auth_url, Some(token_url))
            .authorize_url(CsrfToken::new_random)
            .add_scope(Scope::new(scope_profile.to_string()))
            .add_scope(Scope::new(scope_email.to_string())).url();
            ```

**Comment by maxcountryman** [comment, marked as answer]
From that code snippet it looks like the `client` variable is being assigned to the value of `authorize_url` after adding the scopes. I suspect that's not what you want. What i've done previously is [define a method](https://github.com/maxcountryman/axum-login/blob/1a28cff7b2ecac714535b908a3bd38a7f60ab4fb/examples/oauth2/src/users.rs#L79-L81) to return my authorize URL. This method can be updated to use the `add_scope` invocations to ensure the desired scope is provided.

**Reply by HosMercury** [reply]
Forget my garbage code 
Just tell me where to add scopes according to your code

**Reply by maxcountryman** [reply]
You can add it in that method I linked. I've previously done this with Discord, for example.
